pkg/operator2/routercerts/controller.go

```diff
@@ -0,0 +1,251 @@
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoIngressConfig",
+			Message: fmt.Sprintf("ingresses.config.openshift.io/cluster: %v", err),
+		}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// check domain
+	if len(ingress.Spec.Domain) == 0 {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoIngressDomain",
+			Message: "ingresses.config.openshift.io/cluster: no spec.domain specified",
+		}, nil
+	}
+
+	// get router certs secret
+	secret, err := c.secretLister.Get(c.secretName)
+	if errors.IsNotFound(err) {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoRouterCertSecret",
+			Message: fmt.Sprintf("secret/%v: %v", c.secretName, err),
+		}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// cert data should exist
+	data, ok := secret.Data[ingress.Spec.Domain]
+	if !ok {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "MalformedRouterCertsPEM",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v]: not found", c.secretName, ingress.Spec.Domain),
+		}, nil
+	}
+
+	// certificates should be parse-able
+	certificates, err := crypto.CertsFromPEM(data)
+	if err != nil {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "MalformedRouterCertsPEM",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v] certificates could not be parsed: %v", c.secretName, ingress.Spec.Domain, err),
+		}, nil
+	}
+
+	// categorize certificates
+	var serverCerts []*x509.Certificate
+	intermediateCerts := x509.NewCertPool()
+	rootCerts, err := x509.SystemCertPool()
+	if err != nil {
+		klog.Infof("system cert pool not available: %v", err)
+		rootCerts = x509.NewCertPool()
+	}
+	for _, certificate := range certificates {
+		switch {
+		case certificate.IsCA && bytes.Equal(certificate.RawSubject, certificate.RawIssuer):
+			rootCerts.AddCert(certificate)
+		case certificate.IsCA:
+			intermediateCerts.AddCert(certificate)
+		default:
+			serverCerts = append(serverCerts, certificate)
+		}
+	}
+
+	if len(rootCerts.Subjects()) == 0 {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoRootCARouterCerts",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v] no root CA certificates found in secret, system or proxy", c.secretName, ingress.Spec.Domain),
+		}, nil
+	}
+
+	if len(serverCerts) == 0 {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoServerCertRouterCerts",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v] no server certificates found", c.secretName, ingress.Spec.Domain),
+		}, nil
+	}
+
+	// verify certificate chain
+	for _, serverCert := range serverCerts {
+		_, err = serverCert.Verify(x509.VerifyOptions{
+			DNSName:       c.routeName + "." + ingress.Spec.Domain,
+			Intermediates: intermediateCerts,
+			Roots:         rootCerts,
+		})
+		if err == nil {
```

Comment:
not a fan of setting errs out of current scope and then checking on outer scope.  breaking this out into a function would make it easier to follow

Reply:
Done